### PR TITLE
Add support for typing.Literal in Mapped

### LINKED ISSUE #9187: annotated declarative mis-interprets Literal within Mapped[]
### Describe the bug

I get an `sqlalchemy.exc.ArgumentError: Could not resolve all types within mapped annotation: "Mapped[Status]"` error when I use `from __future__ import annotations` together with a `Mapped` column of type `typing.Literal`.

When I remove `from __future__ import annotations`, everything works as expected.

I'm aware that sqlalchemy already includes built-in support for `enum.Enum`. I'm aware that this is another workaround.

### To Reproduce

```python
# NOTE: Code works if we remove the following line
from __future__ import annotations

from typing import Any, Literal, Optional, get_args

import sqlalchemy.types as types
from sqlalchemy import Dialect
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class LiteralSqlType(types.TypeDecorator):
    impl = types.String
    cache_ok = True

    def __init__(self, literal_type: Any) -> None:
        super().__init__()
        self._possible_values = get_args(literal_type)

    def process_bind_param(
        self, value: Optional[str], dialect: Dialect
    ) -> Optional[str]:
        if value not in self._possible_values:
            raise ValueError(
                f"Invalid literal value '{value}'. Value must be one of"
                f" {self._possible_values}."
            )
        return value


Status = Literal["to-do", "in-progress", "done"]


class Base(DeclarativeBase):
    pass


class Foo(Base):
    __tablename__ = "foo"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[Status] = mapped_column(None, LiteralSqlType(Status))
```

### Error

```
Traceback (most recent call last):
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 189, in eval_expression
    annotation = eval(expression, base_globals, None)
  File "<string>", line 1, in <module>
NameError: name 'to' is not defined

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 1388, in _extract_mappable_attributes
    value.declarative_scan(
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/properties.py", line 681, in declarative_scan
    self._init_column_for_annotation(
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/properties.py", line 717, in _init_column_for_annotation
    argument = de_stringify_annotation(
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 154, in de_stringify_annotation
    elements = tuple(
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 155, in <genexpr>
    de_stringify_annotation(
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 151, in de_stringify_annotation
    annotation = eval_expression(annotation, originating_module)
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/util/typing.py", line 191, in eval_expression
    raise NameError(
NameError: Could not de-stringify annotation 'to-do'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/sqlalchemy_issue.py", line 37, in <module>
    class Foo(Base):
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_api.py", line 768, in __init_subclass__
    _as_declarative(cls._sa_registry, cls, cls.__dict__)
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 246, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 327, in setup_mapping
    return _ClassScanMapperConfig(
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 569, in __init__
    self._extract_mappable_attributes()
  File "/sbt-yocto/build-bactobox/workspace/sources/gmdb/python/.venv/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 1400, in _extract_mappable_attributes
    raise exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Could not resolve all types within mapped annotation: "Mapped[Status]".  Ensure all types are written correctly and are imported within the module in use.
```

### Versions

- OS: `Linux 6f5cb13b6c98 5.15.0-46-generic #49~20.04.1-Ubuntu SMP Thu Aug 4 19:15:44 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
- Python: `Python 3.10.9`
- SQLAlchemy: `sqlalchemy 2.0.0` (as released a couple of days ago)
- Database: N/A
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): N/A

### Additional context

# My understanding of the issue

With `from __future__ import annotations`, python automatically converts all annotations to strings (simplified explanation). This makes it easier to use types out of declaration order. In turn, libraries that use annotations must "de-stringify" the annotations first. This last part is where I think the bug is at.

I think that sqlalchemy is too aggressive with it's "de-stringify annotations" logic.
First, sqlalchemy de-stringifies the literal itself: `'Literal["to-do", "in-progress", "done"]'` becomes `Literal["to-do", "in-progress", "done"]`. So far so good!
Second, sqlalchemy notices that `Literal` is generic (has arguments) and therefore attempts to recursively de-stringify each argument ("to-do", "in-progress", and "done"). However, these arguments are *not* annotations but mere strings. This causes the error.

# Patch to fix the issue

Change the `de_stringify_annotation` function in `sqlalchemy/util/typing.py` so that it explicitly accounts for the `typing.Literal` case:
Specifically, change the recursive part from:

```python
    if include_generic and is_generic(annotation):
        elements = tuple(
            de_stringify_annotation(
                cls,
                elem,
                originating_module,
                str_cleanup_fn=str_cleanup_fn,
                include_generic=include_generic,
            )
            for elem in annotation.__args__
        )

        return _copy_generic_annotation_with(annotation, elements)
    return annotation  # type: ignore
```

to:

```python
    is_literal = get_origin(annotation) is Literal
    if include_generic and is_generic(annotation) and not is_literal:
        elements = tuple(
            de_stringify_annotation(
                cls,
                elem,
                originating_module,
                str_cleanup_fn=str_cleanup_fn,
                include_generic=include_generic,
            )
            for elem in annotation.__args__
        )

        return _copy_generic_annotation_with(annotation, elements)
    return annotation  # type: ignore
```

The code works as expected (no errors) with this fix applied.

Let me know if you have any questions.

*Side note*: Version 2.0 of sqlalchemy is amazing. Great job!

### PR DESCRIPTION
**Pull request by frederikaalund**
Fixes: #9187

See said issue for details.

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by frederikaalund** [comment]
Let me know if you want something changed.

I wanted to split the tests into three distinct tests, but then I would have to copy-paste the `LiteralSqlType` class. Therefore, I chose to only make a single test function instead of three. Let me know if you want it done differently.

I can read from the comments in the test code that you auto-generate `from future import __annotations__` tests based on the contents of the `test_typed_mapping.py` file. Therefore, I only added tests to this file (and I assume that the auto-generation logic does the rest). Let me know if I'm mistaken in my assumptions.

---
In #9187, zzzeek commented:

> we should have tests that also feature a fixed Literal within type_annotation_map

To account for this case, I also changed the `decl_base.py` file. There is a surgical `if ... and not is_literal(...)` change there to make it work.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 7d13f705307bf62560fc831f6f049a425d411374 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 7d13f705307bf62560fc831f6f049a425d411374: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4406

**Comment by zzzeek** [comment]
the changes are fine, I will adjust the test for some conventions we have for that, but basic idea is all there thanks!

this will get merged from gerrit so you will see the PR just closed, but with a happy message

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

thanks for the PR!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4406

File: test/orm/declarative/test_typed_mapping.py [hunk 1]
```diff
@@ -1340,6 +1343,58 @@ class MyClass(Base):
         is_true(isinstance(MyClass.__table__.c.data.type, String))
         eq_(MyClass.__table__.c.data.type.length, 42)
 
+    def test_string_literal(self, decl_base):
```
[the remark this answers is not in the document]
Comment (zzzeek) [reply test/orm/declarative/test_typed_mapping.py:1346]: hadnt touched it yet, had about 9 other reviews today

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

it seemed obvious that one would want ``Literal[]`` with strings to be used as an alternative to ``enum.Enum`` to link to a plain ``Enum`` datatype, and I think it might even be more popular as it's less burdensome, so I implemented that here as well.

this makes for a lot of testing variation since there are at least these combinations:

1. Literal linked to Enum by default without changing the type map
2. Literal with non-strings has to raise an error
3. Literal linked to Enum but we added additional arguemnts to the Enum
4. Literal linked to other kinds of datatypes

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4406

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

yes, lots of combinations

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4406

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4406 has been **merged**. Congratulations! :)

**Pull request closed without merging**